core/templates/dev/head/domain/statistics/FeedbackImprovementTaskObjectFactory.ts

```diff
@@ -101,14 +101,12 @@ angular.module('oppia').factory('FeedbackImprovementTaskObjectFactory', [
        */
       fetchTasks: function() {
         var createNew = this.createNew;
-        return ThreadDataService.fetchThreads().then(function() {
-          return $q.all(
-            ThreadDataService.data.feedbackThreads.map(function(feedback) {
-              return ThreadDataService.fetchMessages(feedback.threadId);
-            }));
-        }).then(function() {
-          return ThreadDataService.getData().feedbackThreads.map(createNew);
-        });
+        return ThreadDataService.fetchThreads()
+          .then(threadData => $q.all([
+            $q.resolve(threadData),
+            ...threadData.feedbackThreads.map(
+              thread => ThreadDataService.fetchMessages(thread.threadId))]))
+          .then(results => results.shift().feedbackThreads.map(createNew));
```

Comment:
I'm a bit confused, reading this code. Why are we using shift() instead of results[0], and why are we singling out the first element in the array?

(Ditto below.)

Reply:
I've tried to clean it up, hows that?